pandas/core/strings.py

```diff
@@ -1105,12 +1105,17 @@ def str_join(arr, sep):
     Returns
     -------
     Series/Index: object
+        The list entries concatenated by intervening occurrences of the
+        delimiter.
 
     Notes
     -----
-    If any of the lists does not contain string objects the result of the join
+    If any of the list items is not a string object, the result of the join
     will be `NaN`.
 
+    If the Series does not contain string objects, performing a join will raise
+    an AttributeError.
```

Comment:
I think that's not very clear. Do you mean that if the Series elements are not lists or `NaN` the join operation will raise an AttributeError?

Reply:
Hmm, I agree. How about:
"If the _supplied_ Series does not contain string objects, performing a join will raise an AttributeError.

```
s = pd.Series([1.1, np.nan]) # applying str.join raises an AttributeError
s = pd.Series([np.nan, np.nan]) # ditto
s = pd.Series([1.1, np.nan, ['swan', 'fish']]) # applying str.join produces something.
```

Reply:
To be accurate, I'd say something like `If the supplied Series does not contain strings or lists, performing a join will raise an AttributeError`.

```
>>> import pandas
>>> pandas.Series(['foo', 'bar']).str.join('')
0    foo
1    bar
dtype: object
```

And if you think it's useful, you can add an example with a case when it raises the exception, in the `Examples` section.

Reply:
👌 We asked the question of mentioning exceptions last night - of the two examples below, which is the preferred method here?

- **Notes** & **Examples**: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.select_dtypes.html
- **Raises** section: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.update.html

Appreciate all your guidance on this, I'm learning a lot! Thank you.

Reply:
We don't pay so much attention to the `Raises` section, to not add extra complexity, but it's surely a nice to have. So feel free to add it.

`Notes` is mainly used for implementation details. In this cases `Raises` can be more appropriate. I'd use `Notes` for example to explain that adding a `NaN` to a `int` column will convert it to `float`, and this kind of stuff.

`Examples` are usually the best way to illustrate things. But we don't want to get crazy with them either. Specially in docstrings, that being mixed with the code, can be annoying if too long. In this case, feel free to choose whether you think it's useful to show with an example a case where the `AttributeError` is raised. Or to omit it if you don't think it add much value. Up to you.